R/utilities-sensitivity-analysis.R

```diff
@@ -618,31 +656,104 @@ plotPopulationSensitivity <- function(structureSets,
     }
   }
 
-
   # allPopsDf is a dataframe that holds the results of all sensitivity analyses for all populations
   allPopsDf <- do.call("rbind", dFList)
 
-  plotList <- list()
 
-  # uniqueQuantitiesAndPKParameters is a dataframe where each row carries a unique combination of QuantityPath and PKParameter
+  # add any missing sensitivity results omitted when applying threshold in getPopSensDfForPkAndOutput
+
+  # get set of unique combinations of outputs and pkParameters.  A plot is built for each combination.
   uniqueQuantitiesAndPKParameters <- unique(allPopsDf[, c("QuantityPath", "PKParameter")])
+
+
+  for (n in 1:nrow(uniqueQuantitiesAndPKParameters)) {
+    op <- uniqueQuantitiesAndPKParameters$QuantityPath[n]
+    pk <- uniqueQuantitiesAndPKParameters$PKParameter[n]
+
+    sensitivityThisOpPk <- allPopsDf[ (allPopsDf$QuantityPath %in% op) & (allPopsDf$PKParameter %in% pk), ]
+
+    # get list of all perturbation parameters used in this plot
+    allParamsForThisOpPk <- unique(sensitivityThisOpPk$Parameter)
+
+    # get the sensitivity results dataframe for this combination of output and pkParameter
+    individualCombinationsThisOpPk <- unique(sensitivityThisOpPk[, c("Quantile", "individualId", "Population")])
+
+    # loop thru each individual in current combination of output and pkParameter
+    for (m in 1:nrow(individualCombinationsThisOpPk)) {
+      qu <- individualCombinationsThisOpPk$Quantile[m]
+      id <- individualCombinationsThisOpPk$individualId[m]
+      pop <- individualCombinationsThisOpPk$Population[m]
+
+      # get list of all perturbation parameters for this one individual that are used in the plot for this combination of output and pkParameter
+      allParamsForThisIndividual <- unique(sensitivityThisOpPk[ (sensitivityThisOpPk$Quantile %in% qu) & (sensitivityThisOpPk$individualId %in% id) & (sensitivityThisOpPk$Population %in% pop), ]$Parameter)
+
+      # create list of the parameters missing for this individual but otherwise shown in this plot for this combination of output and pkParameter
+      missingParameters <- setdiff(allParamsForThisOpPk, allParamsForThisIndividual)
+
+      # loop thru the missing parameters for the current individual
+      for (parNumber in seq_along(missingParameters)) {
+
+        # load the index file of SA results for this individual's population to get the name of the individuals's sensitivity result file
+        indx <- read.csv(saResultIndexFiles[[pop]])
+
+        # get the name of the individuals's sensitivity result file
+        saResFileName <- indx[ (indx$Outputs %in% op) & (indx$pkParameters %in% pk) & (indx$Quantile %in% qu), ]$Filename
+
+        # import SA results for individual
+        individualSAResults <- ospsuite::importSensitivityAnalysisResultsFromCSV(
+          simulation = simulationList[[pop]],
+          filePaths = file.path(structureSet$workflowFolder, structureSet$sensitivityAnalysisResultsFolder, saResFileName)
+        )

```

Comment:
I think you can move these lines in the previous loop and save a few calls of these functions

Reply:
Fixed